tree/tree/src/TTreeCacheUnzip.cxx

```diff
@@ -679,241 +480,379 @@ void TTreeCacheUnzip::ResetCache()
+            // If the block is ready we get it immediately.
+            // And also we don't have to alloc the blks. This is supposed to be
+            // the main thread of the app.
+            if ((fUnzipStatus[seekidx] == 2) && (fUnzipChunks[seekidx]) && (fUnzipLen[seekidx] > 0)) {
 
                //if (gDebug > 0)
-               //   Info("GetUnzipBuffer", "++++++++++++++++++++ CacheLateHIT Block wanted: %d  len:%d fNseek:%d", seekidx, fUnzipLen[seekidx], fNseek);
+               //   Info("GetUnzipBuffer", "++++++++++++++++++++ CacheHIT Block wanted: %d  len:%d req_len:%d fNseek:%d", seekidx, fUnzipLen[seekidx], len,  fNseek);
 
                if(!(*buf)) {
                   *buf = fUnzipChunks[seekidx];
                   fUnzipChunks[seekidx] = 0;
-                  fTotalUnzipBytes -= fUnzipLen[seekidx];
-                  SendUnzipStartSignal(kFALSE);
                   *free = kTRUE;
                }
                else {
                   memcpy(*buf, fUnzipChunks[seekidx], fUnzipLen[seekidx]);
                   delete fUnzipChunks[seekidx];
-                  fTotalUnzipBytes -= fUnzipLen[seekidx];
                   fUnzipChunks[seekidx] = 0;
-                  SendUnzipStartSignal(kFALSE);
                   *free = kFALSE;
                }
 
-               fNStalls++;
+               fNFound++;
 
                return fUnzipLen[seekidx];
             }
-            else {
-               // This is a complete miss. We want to avoid the threads
-               // to try unzipping this block in the future.
-               fUnzipStatus[seekidx] = 2;
-               fUnzipChunks[seekidx] = 0;
 
-               if ((fTotalUnzipBytes < fUnzipBufferSize) && fBlocksToGo)
-                  SendUnzipStartSignal(kFALSE);
+            // If the requested basket is being unzipped by a background task, we try to steal a blk to unzip.
+            Int_t reqi = -1;
+            
+            if (fUnzipStatus[seekidx] == 1) {
+               if (fBlocksToGo > 0) {
+                  for (Int_t ii = 0; ii < fNseek; ++ii) {
+                     Int_t idx = (seekidx + 1 + ii) % fNseek;
+                     if (!fUnzipStatus[idx]) {
+                        Byte_t oldV = 0;
+                        Byte_t newV = 1;
+                        if(fUnzipStatus[idx].compare_exchange_weak(oldV, newV, std::memory_order_release, std::memory_order_relaxed)) {
+                           reqi = idx;
+                           break;
+                        }
+                     }
+                  }
+                  if (reqi < 0) {
+                     fBlocksToGo = 0;
+                  } else {
+                     Int_t locbuffsz = 16384;
+                     char *locbuff = new char[16384];
+                     UnzipCache(reqi, locbuffsz, locbuff);
+                  }
+               } else {
+                  usleep(200000);
```

Comment:
Leave a `TODO` to implement a task-stealing scheme instead of sleeping for a random amount of time.